doc/OSAL-Configuration-Guide.md

```diff
@@ -499,360 +450,263 @@ Using the host vxWorks tools, the path to the same file would be:
+suite of black box unit tests located in the /src/unit-tests directory.
+
+Additionally, OSAL provides a framework for testing called "UT assert".
+The test framework actually uses some of the OSAL BSP facilities to
+execute the tests in a platform-independent manner.
+
+Unit tests are built based on the `ENABLE_UNIT_TESTS` CMake configuration
+variable.  When this is set to TRUE, extra test programs are built to
+validate all of OSAL functionality.
+
+The "black box" functional tests under /src/tests and /src/unit-tests
+operate with an unmodified OSAL library, using a real instance of the OSAL
+to run the tests.
+
+The "white box" coverage test under /src/unit-test-coverage does not use
+OSAL, it only uses the BSP for start up and low-level console access,
+combined with UT assert for the basic framework.  In this testing mode,
+each OSAL subsystem is specially compiled with special flags to enable
+profiling (to track line and branch coverage) and linked with _stub_
+implementations of all other dependencies.  This way, test cases can
+trigger error paths from library calls that would otherwise not be possible
+in a black box test enviornment.
+
+These test programs are executed just as any other application program.
+
+Complete coverage testing is only available for the "shared" layer and
+the "VxWorks" low level implementation layer.  However, because all platforms
+share the same "shared" layer, this coverage test applies to all platforms.
+The low level (OS-specific) implementation layers should be only to translate
+the OSAL calls to the specific C library/OS API provided by that platform,
+while all basic state/table maintenance and error checking is performed at
+the shared layer.
+
+**Note**: The coverage tests are built for the same platform as all other
+applications.  Notably, because the coverage test does not actually use
+the facilities provided by the OS (all OS calls are replaced with stubs),
+the OS-specific coverage test can execute on any platform, it is not limited
+to running on the platform that it is abstracting.  As a result, the VxWorks
+coverage test is typically executed on a Linux host.
 
 ## How to Run the OSAL Unit Tests on the PC / Linux Platform
 
 To build and run the OSAL unit tests, run the following commands from
-the "build" directory: 
+the "build" directory:
 
 #### Build Commands
 
 |  **Shell command**      |    **Description**|
 |---|---|
-|  **\$ cd build**|Change to the build directory.|
-| **\$ make clean-unit-tests**|   Clean the OSAL Core, and all Applications|   
-|**\$ make config**|Copy the osconfig.h for the BSP to the build directory|  
-| **\$ make unit-tests**|Build all of the unit tests.| 
-|  **\$ make gcov**|Run the unit tests, collecting code coverage information.| 
+|  `$ cd build/unit-test-coverage` | Change to the build directory containing coverage tests |
+| `$ make` | Build the OSAL Core, and all tests |
+| `$ make test` | Execute all of the tests programs |
```

Comment:
Could use the top level makefile now to build and report coverage...

Reply:
Revised to show this as two options, first executing a single test program directly from the build tree (useful for debugging in gdb or IDE) and second option to run tests as a batch via "make test" using wrapper makefile.